### ISSUE TITLE
`Local Instance` behaves differently from `Local Definition`

### ISSUE THREAD

**Issue by robbertkrebbers** [coq/coq#13153]
For type class instances, there are two kinds of locality: 1.) locality of namespacing, and 2.) locality of the hint. Both the `Local` keyword and the `#[local]` appear to control the latter in case of an `Instance`, but the former in case of a `Definition`. It should be possibly to control both.

Here's an example:

```coq
Class C (n : nat) := {}.

Module Import foo.
  Local Instance c9 : C 9 := {}.
  #[local] Instance c10 : C 10 := {}.
End foo.

Check c9. (* works *)
Check c10. (* works *)
Check (_ : C 9). (* fails *)
Check (_ : C 10). (* fails *)
```

Note that this issue is in fact even more complicated, for locality of the hint., there are actually three options: a.) the hint is local to the secion 2.) the hint is local to the module/file 3.) the hint is properly global. I don't think there's a proper way to control that.

/cc @RalfJung as this came up in some discussions about the use of `Local` and `#[local]` in Iris.

**Comment by Zimmi48** [comment]
Yes, `Local` and `#[ local ]` are strict synonyms but are interpreted per command. In the case you raise, it is indeed ambiguous what it means, and we should be able to distinguish the two by providing different attributes. In the meantime, you can work around this issue by using `Local Definition` followed by `Local Existing Instance`:

```coq
Module Import foo.
  Local Definition c9 : C 9 := Build_C _.
  Local Existing Instance c9.
End foo.
```

**Comment by RalfJung** [comment]
I recently learned that `Local Ltac` uses yet another way to interpret the `Local` keyword: the tactic becomes entirely inaccessible outside the current module.

So, we have one keyword doing three completely different things, and the effect of `Local Definition` is entirely unavailable for `Ltac` I think.

**Comment by ppedrot** [comment]
`Local Definition` is essentially a hack that is a trade-off between abstraction and usability. The only sane way to fully hide a definition in CIC is to go through a module type but this is unwieldy from the user point of view, and problematic from a kernel point of view (you know, modules!). So I wouldn't pick the local semantics of `Definition` as the standard one.

**Comment by Zimmi48** [comment]
@RalfJung Actually, that's the most frequent meaning: `Local` restricts things to the current module (or section). Because it is not possible in practice for definitions (since they could appear in the type of other constants), they are masked as well as we can by making them only accessible by their qualified name. See the documentation at: https://coq.inria.fr/refman/proof-engine/vernacular-commands.html#controlling-the-locality-of-commands

**Comment by RalfJung** [comment]
@ppedrot `Local Definition` is extremely useful though, at least so far it is the best way I have figured out to do "library design" in Coq without enormous overhead. (I consider Module Type to have too much overhead because one has to state the public interface twice.) So in that sense it would be nice to be able to do the same with tactics that are part of a module. Currently there is no way to have consistent naming/scoping for a `Local Definition` and an accompanying `Ltac`.

**Comment by ppedrot** [comment]
> Local Definition is extremely useful though

I know, I am the one who implemented it...

**Comment by Zimmi48** [comment]
I was not aware that there could be a use case for this behavior for something that you can fully mask. Since it's actually useful, what about introducing a new attribute, e.g. `qualified` that would encompass the current behavior of `Local Definition`. We could keep `Local Definition` working for backward-compatibility (if deprecating is viewed as creating too much churn) but we would generally recommend the use of the `qualified` attribute instead (both on `Definition` and on other commands declaring global objects).

**Comment by RalfJung** [comment]
So `#[qualified] Local Instance` would also make sure that the instance is only locally registered *and* the name is only available fully qualified? (Seems somewhat tedious to have to state both of these attributes but that's probably better than magically making one imply the other.)

**Comment by Zimmi48** [comment]
Correct.

**Comment by Zimmi48** [comment]
In #13831 (and following the last Coq Call https://github.com/coq/coq/wiki/Coq-Call-2021-02-03), I'm adding proper documentation for locality attributes `local` and `global` and a warning that the semantics is different for `Definition`, `Theorem` and variants. But the next step is definitely to introduce the `qualified` attribute and probably to deprecate `local` on `Definition`, `Theorem` and variants in favor of `qualified`.
